working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+Javascript object, providing a wrapperless interoperation capability.
+
+
+```dart
+extension struct Window(JSObject o) {
+  // Signatures provided here for methods to be delegated to the underlying
+  // JSObject
+  external bool get closed;
+  // etc
+}
+
+void test(Window w) {
+  // Window methods can be called using Dart syntax
+  if (w.closed) {...}
+
+  // Windows are represented as the underlying object
+  assert(w is JSObject);
+
+  // Windows can be cast to the underlying object
+  JSObject o = w as JSObject;
+
+  // Extension struct types are reified as the underlying representation type.
+  List<Window> l = [w];
+  assert(l is List<JSObject>);
+}
+```
+
+Extension structs are also useful for providing a lightweight facade over an
+existing type.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) {
+  Nat(int x) : assert(x >= 0), _x = x;
+  Nat.zero() : _x = 0;
+
+  Nat get succ => Nat(_x+1);
+  Nat plus(Nat other) => Nat._x + other._x;
+
+  // Override the underlying isNegative operation (incorrectly)
+  bool get isNegative => true;
+}
+
+void test() {
+  var n1 = Nat(3);
+  var n2 = Nat.zero();
+  assert(n2.succ.succ.succ == n2);
+
+  //The underlying representation is still as an int
+  assert(n1 is int);
+
+  // The static type is used to dispatch the method calls
+  assert(n1.isNegative);
+
+  // If the static type is lost, dispatch goes to the underlying object.
+  dynamic d = n1;
+  assert(!d.isNegative);
+}
+```
+
+Extension structs may delegate members to the underlying field.  They may also
+implement interfaces, but only if the underlying field implements the interface.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) implements Comparable<num> {
+  // Constructors etc as above
+
+
+  bool get isEven;  // Abstract definition delegates to _x.isEven
+}
+
+void test() {
+  var n = Nat(2);
+  // Same as 2.isEven
+  assert(n.isEven);
+  // Comparable interface allows access to the Comparable methods on int
+  assert(n.compareTo(2) == 0);
+
+  // Since Nat implements Comparable<num>, it may be assigned to it
+  Comparable<num> c = n;
+  // The representation is still as an integer
+  assert(c is int);
+}
+```
+
+Extension structs have no inheritance.
+
+```dart
+// Static error
+extension struct PositiveNumber(int _x) extends Nat {...}
+```
+
+Extension structs do not define a signature, and hence cannot be implemented.
+
+```dart
+// Static error
+extension struct AlternativeNat(int _x) implements Nat { ...}
+
+// Static error
+class MockNat implements Nat {...}
+```
+
+
+## Extension structs in more detail
+
+
+### Introduction form
+
+Extension structs are defined by adding the keyword "extension" before a struct
+definition.  The syntax for extension structs is otherwise exactly identical to
+that of general structs.  However, extension structs are subject to additional
+restrictions.  The most important of these is that extension structs may only
+contain a single entry in their primary constructor.
```

Comment:
If I consider this proposal in isolation, I buy this restriction. However, what if we consider this together with the proposal on records, could we treat more than one entry in the constructor as a record (rather than multiple fields?) Even in the context of regular structs, is the distinction of 3 fields vs 1 record field relevant anymore?

Reply:
Yes.  I hesitate about it a bit though.  There are three interpretations I considered for this:
  - `extension struct Foo(int x, int y)` is represented as a two-tuple.  When you call the constructor, there is an auto-boxing into the tuple representation.
    - I don't like this because auto-boxing.  I'd at least like something there as an opt in.
  - `extension struct Foo(int x, int y)` is represented as a two-tuple, and the constructor has a single tuple parameter
    - the boxing is still there, but at least the user knows it
    - the syntax ends up bleh `var x = Foo((3, 4))`
  - `extension struct Foo(int x, int y)` is represented as two objects, not one.  It is not a subtype of object.  You cannot use it polymorphically.  You can have variables of type Foo, and they have move semantics.
    - This is intriguing - I like the new functionality it gives you
    - This is weird, and it may surprise people writing APIs (they may innocently add a field, and be surprised that weird things happen)

Reply:
Filed https://github.com/dart-lang/language/issues/2371 to discuss this.